dev-docs/patches.md

```diff
@@ -0,0 +1,95 @@
+# Working with patches
+
+## Patches in Nix
+
+There are two common ways to apply patches in Nix:
+
+- Use the `patches` attribute of a builder ([manual](https://nixos.org/manual/nixpkgs/unstable/#var-stdenv-patches)):
+
+  ```nix
+  stdenv.mkDerivation {
+    # ...
+    patches = [
+        ./0001-first-change.patch
+        ./0002-another-change.patch
+    ]
+    # ...
+  }
+  ```
+  This is most commonly used and can be encountered nearly everywhere upstream in nixpkgs.
+
+- Use the `applyPatches` function ([noogle](https://noogle.dev/f/pkgs/applyPatches)) to patch an existing source and derive a new source:
+
+  ```nix
+  stdenv.mkDerivation {
+    # ...
+    src = applyPatches {
+        src = fetchFromGuiHub { /* ... */ }
+    patches = [
+        ./0001-first-change.patch
+        ./0002-another-change.patch
+    ]
+    # ...
+  }
+  ```
+
+  This is useful in situations where the patches sources should also be used by other consumers, as the
+  patched source can inherited, but `patches` aren't part of the output set.
+
+Patches can either be consumed from a local file or pulled from remote:
+
+```nix
+patches = [
+  (fetchpatch {
+    url = "https://github.com/...";
+    stripLen = 2; # Sometimes the patch strip level must be configured
+    hash = "sha256-AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA=";
+  })
+];
+```
+
+If there aren't any changes needed to an upstream patch, it's good practice to fetch the patch from remote. If the changes to upstream patches are needed or patches aren't available upstream, the
```

Comment:
I'd argue that a patch needs to be in some reliable branch (main, release/xxx) to qualify for remote inclusion, to minimize the chance of losing it due to garbage collection etc.

Reply:
While I see the point of garbage collection on the GitHub side, I don't think this is too big of a problem. From my experience they don't run aggressive gc on repos and commit stay around for quite some time. The reason I'd rather not restrict this further is that it is quite common to reference a commit from an open PR that we've just opened, which is ok imo.

Actually did some research, seems like at least in some scenarios dangling commits are kept around for quite some time: https://stackoverflow.com/questions/4367977/how-to-remove-a-dangling-commit-from-github

Reply:
I like how it's phrased now, and thanks for doing the research.